### PR TITLE
fix(core): moving unhandled error throw

### PR DESCRIPTION
**Pull request by thejuan**
What does this implement/fix? Explain your changes.
---------------------------------------------------

When an error was unhandled by the error middleware, it was being re-thrown and added to itself. 
This was causing issues with downstream testing libraries which JSON.stringify the error and blow up as it's a circular dep.
Polluting logging having duplicate errors inside itself.

Does this close any currently open issues?
------------------------------------------
No

Where has this been tested?
---------------------------

CI/CD

Todo list
---------

[x] Feature/Fix fully implemented
[TODO ] Added tests

### CODE REVIEW

**Comment by thejuan** [comment]
Any guide for running tests locally? `ava` seems to barf on raw js

**Comment by willfarrell** [comment]
Thanks for the PR. This would be a breaking change. So may need to go in v3, if approved.

Can you write a failing example to show the edge case you're describing?

**Comment by willfarrell** [comment]
> Any guide for running tests locally? ava seems to barf on raw js 
I just run `npm run test:packages:unit` or `PACKAGE=core npm run test:unit` locally without needing to build just fine

**Comment by thejuan** [comment]
Thanks will do, what does it break?

**Comment by thejuan** [comment]
> npm run test:packages:unit

```
npm run test:packages:unit

> middy-monorepo@2.4.1 test:packages:unit /home/adam/hatch/libs/middy
> lerna exec --bail --concurrency 5 npm run test:unit

lerna notice cli v4.0.0
lerna info Executing command in 28 packages: "npm run test:unit"

> @middy/core@2.4.1 test:unit /home/adam/hatch/libs/middy/packages/core
> ava


  ✖ No tests found in __tests__/index.js

  ─

  Uncaught exception in __tests__/index.js

  __tests__/index.js:3

   2: const sinon = require('sinon')      
   3: const middy = require('../index.js')
   4:                                     

  /home/adam/hatch/libs/middy/packages/core/index.js:2
    plugin?.beforePrefetch?.()
           ^

  SyntaxError: Unexpected token '.'

  › SyntaxError: Unexpected token '.'
  › Object.<anonymous> (__tests__/index.js:3:15)

  ─

  1 uncaught exception
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! @middy/core@2.4.1 test:unit: `ava`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the @middy/core@2.4.1 test:unit script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/adam/.npm/_logs/2021-06-21T04_31_11_885Z-debug.log
lerna ERR! npm run test:unit exited 1 in '@middy/core'
lerna ERR! npm run test:unit exited 1 in '@middy/core'
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! middy-monorepo@2.4.1 test:packages:unit: `lerna exec --bail --concurrency 5 npm run test:unit`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the middy-monorepo@2.4.1 test:packages:unit script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/adam/.npm/_logs/2021-06-21T04_31_11_931Z-debug.log
```

**Comment by willfarrell** [comment]
Oooh, when running locally you need to be using nodejs v14. The CI transpiles to Node v12 before running.

I consider all changes to `core` breaking until all existing tests are passing. In this case `Should trigger all plugin hooks Rejected promise returned by test` is failing.

**Comment by thejuan** [comment]
Updated tests to show the issue as well as show it's not a breaking change.
(Test handlers were not being initialised correctly)

**Comment by thejuan** [comment]
Fixed logic to work with current tests (it was a bug previously) :)
Should be all good now.

**Comment by thejuan** [comment]
There a number of ways to solve the issue, this was just the first idea.
Could also check if the errors are the same in the re-throw section and throw the original error.
Could always wrap errors in a composite error.

The main issue is that there isn't a property on the error that ref equals the error itself as this blows up `JSON.stringify`

**Comment by willfarrell** [comment]
Thanks for your patience on this one, it's been busy lately. I see what you're talking about, having the same error nested doesn't make sense, for sure a bug. The extra check to see if it's an error isn't necessary if we move the check within the first catch, but after the nested catch. This ensures `requestEnd` truly is the end, and isn't running extra code on successful requests. Can you make this quick change, then I can push a patch release. See below:

```js
...
    request.error = e
    try {
      await runMiddlewares(request, onErrorMiddlewares, plugin)
    } catch (e) {
      // Save error that was interupted before it could be handled
      e.originalError = request.error
      request.error = e
      throw request.error
    }
    // Catch if onError stack hasn't handled the error
    if (request.response === undefined) throw request.error // <-- move here
  } finally {
    await plugin?.requestEnd?.()
  }
  return request.response
}
```

Thanks again for the PR.

**Comment by thejuan** [comment]
Yeah, that makes more sense!
Done.

**Comment by thejuan** [comment]
Will look at tests

File: packages/core/index.js [hunk 1]
```diff
@@ -96,17 +96,19 @@ const runRequest = async (
...
-      if (request.response === undefined) throw request.error
     } catch (e) {
       // Save error that wasn't handled
       e.originalError = request.error
       request.error = e
+
       throw request.error
     }
+    // Catch if onError stack hasn't handled the error
+    if (request.error && request.response === undefined) throw request.error
```
Comment (willfarrell) [inline packages/core/index.js:107]: `request.error && ` is redundant

**Review by willfarrell — CHANGES_REQUESTED** [review]

File: packages/core/index.js (hunk 1 above)
Comment (thejuan) [reply packages/core/index.js:107]: True, the middleware may have changed the error to throw.
So what about `throw request.error || e`

Comment (willfarrell) [reply packages/core/index.js]: I don't expect someone to override the Error to something falsy or nullish (famous last words). You can leave the `|| e` out for now.

**Comment by willfarrell** [comment]
Looks like a failing test was pushed to main, I'll fix that.

File: packages/core/index.js (hunk 1 above)
Comment (thejuan) [reply packages/core/index.js]: Fixed

**Comment by willfarrell** [comment]
Fix pushed

**Review by willfarrell — APPROVED** [review]

**Comment by willfarrell** [comment]
Thanks again for the PR, I'll try and mint a release tomorrow.
